archive/100-theorems-list/37_solution_of_cubic.lean

```diff
@@ -0,0 +1,201 @@
+
+
+noncomputable def ω : ℂ := ⟨-1/2, (real.sqrt 3)/2⟩
+
+lemma omega_cube_eq_one : ω ^ 3 = 1 :=
+begin
+  unfold ω,
+  have h3 : real.sqrt 3 * real.sqrt 3 = 3,
+  { rw [← pow_two (real.sqrt 3), @real.sq_sqrt 3 (by linarith)], },
+  rw [pow_succ, pow_succ, pow_one],
+  ext,
+  iterate 2
+  { simp only [mul_re, mul_im],
+    rw ← sub_eq_zero,
+    field_simp [mul_add, h3],
+    ring }
+end
+
+lemma omega_sum : 1 + ω + ω^2 = 0 :=
+begin
+  have : 1 - ω ≠ 0,
+  { unfold ω, intro h, rw complex.ext_iff at h, norm_num at h },
+  rw [← mul_right_inj' this, mul_zero],
+  calc (1 - ω) * (1 + ω + ω^2)
+        = 1 - ω^3 : by ring
+    ... = 0 : by rw omega_cube_eq_one; norm_num
+end
+
+/-- The solution of the monic cubic equation, with 2nd coefficient set to zero. -/
+lemma cubic_basic_eq_zero_iff
+  (hp_nonzero : p ≠ 0)
+  (hr : r^2 = q^2 + p^3)
+  (hs3 : s^3 = q + r)
+  (ht : t * s = p)
+  (x : ℂ) :
+    x^3 + 3 * p * x - 2 * q = 0 ↔
+      x = s       - t        ∨
+      x = s * ω   - t * ω^2  ∨
+      x = s * ω^2 - t * ω    :=
+begin
+  have h₁ : ∀ x a₁ a₂ a₃ : ℂ, x = a₁ ∨ x = a₂ ∨ x = a₃ ↔ (x - a₁) * (x - a₂) * (x - a₃) = 0,
+  { intros,
+    iterate 2 { rw [mul_eq_zero] },
+    iterate 3 { rw [sub_eq_zero] },
+    rw or.assoc, },
+  rw h₁,
+  suffices : x ^ 3 + 3 * p * x - 2 * q
+    = (x - (s - t)) * (x - (s * ω - t * ω^2)) * (x - (s * ω^2 - t * ω)),
+  { rw this, },
+  have hc : s^3 - t^3 = 2 * q,
+  { have : s ≠ 0 := λ h, by { apply hp_nonzero, rw [h, mul_zero] at ht, exact ht.symm },
+    have h_nonzero: q + r ≠ 0 := by { rw ← hs3, exact pow_ne_zero _ this },
+    have hp3 : p^3 = r^2 - q^2 := by rw [hr]; ring,
+    calc    s^3 - t^3
+          = s^3 - p^3/s^3 : by { rw [← ht], field_simp, ring }
+      ... = (q+r) - (r^2-q^2)/(q+r) : by rw [hs3, hp3]
+      ... = (q+r) + (q-r) * ((q+r) / (q+r)): by ring
+      ... = 2 * q : by rw div_self h_nonzero; ring },
+  symmetry,
+  calc (x - (s - t)) * (x - (s * ω - t * ω^2)) * (x - (s * ω^2 - t * ω))
+      = x^3 - (s-t) * (1+ω+ω^2) * x^2
+        + ((s^2+t^2)*ω*(1+ω+ω^2) - s*t*(-3 + 3*(1+ω+ω^2) + ω*(ω^3-1))) * x
+        - (s^3-t^3)*ω^3 + s*t*(s-t)*ω^2*(1+ω+ω^2) : by ring
+  ... = x^3 + 3*(t*s)*x - (s^3-t^3) : by { rw [omega_sum, omega_cube_eq_one], ring }
+  ... = x^3 + 3*p*x - 2*q : by rw [ht, hc]
+end
+
+/-- the solution of the monic cubic equation. -/
+lemma cubic_monic_eq_zero_iff
+  (hp : p = (3 * c - b^2) / 9)
+  (hp_nonzero : p ≠ 0)
+  (hq :  q = (9 * b * c - 2 * b^3 - 27 * d) / 54)
+  (hr : r^2 = q^2 + p^3)
+  (hs3 : s^3 = q + r)
+  (ht : t * s = p)
+  (x : ℂ) :
+    x^3 + b * x^2 + c * x + d = 0 ↔
+      x = s       - t       - b / 3 ∨
+      x = s * ω   - t * ω^2 - b / 3 ∨
+      x = s * ω^2 - t * ω   - b / 3 :=
+begin
+  let y := x + b / 3,
+  have h₁ : x^3 + b * x^2 + c * x + d = y^3 + 3 * p * y - 2 * q,
+  { dsimp [y],
+    rw [hp, hq],
+    field_simp, ring, },
+  have h₂ := @cubic_basic_eq_zero_iff p q r s t hp_nonzero hr hs3 ht y,
+  rw [h₁, h₂],
+  dsimp [y],
```

Comment:
Non-terminal `dsimp` should be avoided, like `simp`.
```suggestion
  dsimp only [y],
```
You might also be able to `dunfold y` here (it's a flavor of `dsimp`).

Reply:
FYI @kmill , nonterminal `dsimp` isn't as objectionable as `simp`. New `dsimp` lemmas are added very rarely so these proofs are a lot more stable. That said I do try to avoid it myself!